Add auth per email rate limit.

stephank opened the pull request:
First stab at an implementation of #70.

This is the third implementation option mentioned there, with I believe some small improvement. The details are explained here: https://github.com/portier/portier-broker/blob/d2374ae9ac2e0e3e306ce60fb5f666d8975ab406/src/store_limits.rs#L77-L86

callahad commented:
I'm still wrapping my head around this strategy. Questions that come to mind:

1. Is `handlers/oidc.rs` the right place to be testing limits, or do we only care about ratelimiting for code-based (email_bridge.rs) workflows?

2. Should there be a higher level interface to `incr_and_test` to avoid the handler having to generate and submit the current time?

3. Can we avoid passing a reference to the store (`app.store.limits.incr_and_test(&app.store, ...)` feels a bit strange to me -- if it's a method chained down from `app.store`, why do we still have to pass `&app.store` as an argument? Or if we really must pass the reference, why is `limits` a member of `app.store` instead of a standalone function?

stephank commented:
1. Good point. Rate limiting emails is definitely the most important, but shouldn't we also limit sessions in general? A non-browser client could spam `/auth` with a GMail address to flood our Redis. Should it be a separate limit?

2. I originally tried to implement limiting by IP, but then realised we don't have the real IP, because we're behind a proxy. The idea was to measure time once, and use that everywhere. I somehow picked up that `gettimeofday` and friends are slow, and I was unsure what Rust is actually doing, but I'm now thinking it's probably stupid to optimise that in this case. Will change it.

3. This matches the design of `StoreCache`. I'm now thinking these could just as well be regular functions, instead of methods. We don't even need the `StoreCache` and `StoreLimits` structs or `self`. If that's okay, I'll add that change to the PR.

stephank commented:
I've rebased it and tried to address all three points.

Regarding the first, I've split up the limit in an `auth` limit for all kinds, and an `auth_email` for emails specifically.

callahad commented on config.toml.dist:
```diff
@@ -51,3 +51,10 @@ secret =
+# Limit the number of session of any kind.
+auth = ["per-second:3:1:1", "per-minute:20:60:5"]
+# Limit the number of email loop authentications.
+auth_email = ["per-second:1:1:1", "per-minute:10:60:5"]
```
I wonder if the benefits of this degree of granularity are worth the complexity / cognitive overhead?

E.g., Could we automatically generate a reasonable granularity value? Do we need names for each limit, even though they might look like magic strings and confuse people? Should we deliberately turn them into magic strings? Do we need a vector, or could we get by with a single limit for each case?

Could we get by with a single stupid value like this, and would it be better?

```
[smtp]
per_user = "10/min"
```

I genuinely don't know what the best answer is to these would be. I do know that parsing the limits section hurts my head a little more than I'd like, but not enough to block merging of a patch that *does* work, as long as you're generally happy with this design.

onli replied:
Just as a confirmation, when reading through that patch I though the same, that it would be nice to make that easier. I already have issues understanding the bucket system (never encountered that before), adding granularity on top of that in the config does not help.

stephank replied:
I'm starting to think simpler is probably better too.

Calculating granularity is something I tried, but troublesome while we enforce `timespan % granularity == 0`. Something that could be worked around.

I think, as long as we use small windows, something like `10/min` per email with a single key could work just fine.

callahad commented on src/store_limits.rs:
```diff
@@ -0,0 +1,118 @@
+            .hdel(&key_str, hdel_fields).ignore()
+            .expire_at(&key_str, (unixtime + config.timespan) as usize).ignore()
+            .hvals(&key_str)
+            .query(&store.client)?;
```
Would it make sense to try to merge all of the limit checks into a single Redis request? What's the overhead of hitting Redis four times for every email-based auth request? Eh, probably not worth worrying about. :)

stephank replied:
Four times, as in, four limit configs? The overhead of the individual queries in a single pipelined transaction is probably not bad. Not sure what the right tradeoff is otherwise.

callahad commented:
Wait, what about clock skew?

If the Broker is a few seconds behind Redis, we could end up setting an `EXPIREAT` in the past, and never actually enforcing a ratelimit.

If there's more than one Broker, they could disagree about which bucket is active, and could end up clearing buckets that should still be valid.

callahad commented:
Redis does have a `TIME` command that we could use to determine the current server time, but I wonder if we can somehow develop a scheme that doesn't rely on calculating timestamps, and which only uses the offset-based `EXPIRE` command?

callahad commented:
(I'm hesitant to rely on `TIME`, since network I/O could pause indefinitely between receiving the `TIME` response and issuing our subsequent query.)

callahad commented:
I found a few good articles on rate limiting strategies with Redis:

- Introduction to rate limiting with Redis: [Part 1](http://www.dr-josiah.com/2014/11/introduction-to-rate-limiting-with.html), [Part 2](http://www.dr-josiah.com/2014/11/introduction-to-rate-limiting-with_26.html)
- [Better rate limiting with Redis sorted sets](https://engineering.classdojo.com/blog/2015/02/06/rolling-rate-limiter/)

I really like that first article, and it seems to build up to a solution similar to the one in this pull request, and discusses all of the associated pitfalls and corner cases along the way. Ultimately, it ends up with a massive lua script that executes within the Redis server, which works, but seems like overkill. It also still requires two round-trips in order to first get the server's time, and then execute the custom script.

I'm starting to lean back towards a naive, single-bucket-per-email ratelimit, and only limiting the smtp provider. If we're defining our ratelimit in terms of requests per minute, I don't think we need to worry about smoothing across multiple buckets like we would if we were dealing with per-hour or per-day limits.

callahad commented:
Running a parallel branch at https://github.com/portier/portier-broker/compare/devil-in-the-details to see if another strategy is reasonably effective and sufficiently simple...

stephank commented:
I like the simpler strat. Was looking for `script load` in redis-rs, but apparently it'll lazy load scripts for you. Just have to create a `Script` and go. :)

onli commented:
Came here to propose the simpler strategy :) We should have an per-IP-rate-limiting around that, so one or a chosen small number of attackers can't block one email address. Can be on the proxy-level, or we just figure out how to reliably use X-Forwarded-For or another Proxy header iff behind a proxy, using the IP otherwise.
